Fix sizing data for zero load case - ref issue #4893

Nigusse opened the pull request:
PR: 90946736 wshp #4567

rraustad commented:
This seems like a good fix for times when a cooling load does not exist. The change to else if for cooling keeps conditions separate for cooling, heating, and no load conditions. What to initialize the no loads conditions to (or sizing data in general for defaults) is another issue, but for now 0 seems adequate. Another review wouldn't hurt, but this has been initially reviewed and is ready for merge.

mjwitte commented:
@Nigusse @rraustad I'm not following the discussion completely (over in the pivotal ticket, moving the discussion here).  For water coils, if the leaving design supply humidity ratio is greater than the design condition entering humidity ratio, then the leaving humidity ratio is reset to match the entering.  It sounds like the opposite adjustment is being made here?  There are other cases where a dry climate could result in the same condition (design leaving hr > entering hr) even if there is not a zero load.

Nigusse commented:
@mjwitte @rraustad When there is tiny heating design (SysOutputProvided > 0) load during cooling design day calculation (basement zone), then CalcZoneSizing( ControlledZoneNum, CurOverallSimDay ).CoolZoneHumRat is set to zero, hence FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak is set to zero initially.  Later in the code FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak is set to design supply air humidity ratio since its value was set zero during initialization. Depending on the value of SysOutputProvided two zone equipment will be initialized to different HR value.  As the result one of the zone is sized to SHR > 1 and the other zone is sized to 0.0 < SHR< 1. This fix was meant to get at least consistent set of sizing capacities and SHR. The current initialization logic using the "SysOutputProvided" does not capture all cases.

rraustad commented:
The underlying problem is that the sizing arrays get initialized prior to use. In the case of zone hum rat, this is initialized to OA hum rat (Node( ZoneNodeNum ).HumRat = OutHumRat; at line 376 ZoneEquipmentManager). During sizing, the peak conditions are saved for use in the child component. If there was a heating load, the peak heating load conditions are saved, OTHERWISE (i.e., else instead of else if) the conditions are saved in the cooling arrays. This issue showed a cooling coil with SHR>1, and sensible capacity less than total capacity. This was a result of very low zone humidity ratios (since there were no zone cooling loads during sizing) being used to size total cooling capacity while delta T was used for sensible capacity. The correction here is to only save cooling/heating peak conditions if there is a cooling/heating load, and save 0's for no load conditions. Then child component sizing can test for 0 and act accordingly. In the case of this coil, the 0 hum rat triggered using the Sizing:Zone supply air humidity ratio instead of an unbelievably low zone humidity ratio (which was set because of using the else instead of using else if to definitively set the conditions based on load).

Here you can see that the zone sizing data is used when there is no cooling load (the else here):
line 2376 of ZoneEquipmentManager

// initialize sizing conditions if they have not been set (i.e., no corresponding load) to zone condition
if ( FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak > 0.0 ) {
	FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak = min( FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak, PsyWFnTdpPb( FinalZoneSizing( CtrlZoneNum ).ZoneTempAtCoolPeak, StdBaroPress, RoutineName ) );
} else {
	FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak = ZoneSizing( CtrlZoneNum, DDNumF ).CoolDesHumRat;
}

mjwitte commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -640,7 +640,16 @@ namespace ZoneEquipmentManager {
+				CalcZoneSizing( CurOverallSimDay, ControlledZoneNum ).CoolLoad = 0.0;
+				CalcZoneSizing( CurOverallSimDay, ControlledZoneNum ).CoolMassFlow = 0.0;
+				CalcZoneSizing( CurOverallSimDay, ControlledZoneNum ).CoolZoneTemp = Node( ZoneNode ).Temp;
+				CalcZoneSizing( CurOverallSimDay, ControlledZoneNum ).CoolZoneHumRat = ZoneSizingInput( ControlledZoneNum ).CoolDesHumRat;
```
@Nigusse @rraustad This line is failing with an array bounds error (and I suspect line 651 will also) as reported here
https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-f138d068e3cbcaddbdc9d8a5d688de9716021344-x86_64-Linux-Ubuntu-14.04-gcc-4.8-IntegrationCoverage-Debug.html
This was failing before I merged in develop and made the row-major updates.  It concerns me that this spot in the code would be looking back at ZoneSizingInput.  Seems there should be a different place to get this from, but I'm not following the logic here completely.  You can debug this by running just about any example file in the debugger (I ran 5ZoneAirCooled) and it failed on this line.

mjwitte replied:
@Nigusse @rraustad I suspect this may be why the unit test fails on some platforms, but I have not confirmed that.

Nigusse replied:
The original attempt to fix this issue was to set all the eight variables in the "else" case to zero  as follows on lines 644 - 651:
                CalcZoneSizing( ControlledZoneNum, CurOverallSimDay ).HeatLoad = 0.0;
                CalcZoneSizing( ControlledZoneNum, CurOverallSimDay ).HeatMassFlow = 0.0;
                CalcZoneSizing( ControlledZoneNum, CurOverallSimDay ).HeatZoneTemp = 0.0;
                CalcZoneSizing( ControlledZoneNum, CurOverallSimDay ).HeatZoneHumRat = 0.0;
                CalcZoneSizing( ControlledZoneNum, CurOverallSimDay ).CoolLoad = 0.0;
                CalcZoneSizing( ControlledZoneNum, CurOverallSimDay ).CoolMassFlow = 0.0;
                CalcZoneSizing( ControlledZoneNum, CurOverallSimDay ).CoolZoneTemp = 0.0;
                CalcZoneSizing( ControlledZoneNum, CurOverallSimDay ).CoolZoneHumRat = 0.0;

Then these variables are updated or set to the design sizing input in "UpdateZoneSizing" routine. I think, we thought initializing to the design input values early was better.

@mjwitte @rraustad I will test Mike's findings and reset these variables to zero for else case.

Nigusse replied:
@mjwitte I pulled the latest of 90946736 wshp #4567 branch but I see many variables inaccessible error flag in ZoneEquipmentManager code. My be I should start from develop again. Was there any conflict when you merged in develop?

mjwitte commented:
@Nigusse I've compiled the current 90946736 branch, but I did have to delete my cmake cache and regenerate recently (when I was building the HX control branch a day or two ago).  You should not need to restart with develop.

Nigusse commented:
@mjwitte @rraustad I have fixed the allocation issue and tested using an example file.  It run fine. Also I fixed unit test assertion failure identified using the individual unit test.  There was a variable that was not allocated. So, let us see how the CI machine runs.

Myoldmopar commented:
Looks like just regressions, no unit test failures now, @Nigusse.  Nice work!

mjwitte commented:
@Nigusse For PipingSystem_Underground_FHX, the heat pump now sizes about 3x larger, while the other three files that show diffs are sizing smaller, about 1/3x.  Is that what you expect?

Nigusse commented:
@mjwitte @rraustad The unit test for this bug fix has passed. The CI shows diffs in four example files shown below: 
(1) PipingSystem_Underground_FHX.idf
(2) ZoneSysAvailManager.idf
(3) ZoneVSWSHP_wDOAS.idf
(4) ZoneWSHP_wDOAS.idf

The first example file that had total cooling capacity less than the sensible cooling capacity and hence SHR > 1.0. Now the total cooling capacity is greater than that of the sensible and the SHR is 0.67 and the warning error associated is gone. This is expected.

The other three test files show consistent trends, the total cooling capacity in each case reduced by a factor 2 to 3.7 compared to the previous value.  Similarly the sensible capacity and design flow rates reduced proportionally.

Nigusse commented:
@mjwitte For the PipingSystem_Underground_FHX example file the sensible capacity did not change, so the total cooling capacity has to be greater by at least 1.7x to make the SHR =1. Based on the specified sizing inputs and stepping through the code, the sizing is done correctly.  So, the result is as expected. I have to check one example from the other three to see what is happening in the sizing calculation.

Nigusse commented:
@mjwitte @rraustad I tested the three example files (#2 - #4), in all the three cases the capacities were reduced by a factor of 2 to 3.7 range.  Note that the Calculated design cooling load for the zone was zero in all cases.  This capacity reduction happens only when the zone cooling load is zero. At the same time the reduced capacities of the equipments provides enough cooling to the zone, and even at the reduced capacities the PLR of the cooling operation was well below 0.22, and in one of the test files it is below 0.05.  The zone thermostat set point temperature was maintained in all cases. So, I am comfortable even the capacities were reduced by a factor 2 or 3. After all we are sizing the equipment for zero zone cooling load. Does this really matter?

mjwitte commented:
@Nigusse Ran a full regression test suite with annual simulations and daily output.  Found big diffs in the output for the four test cases already discussed above.  But there are about 100 test cases where there are big diffs in the ssz output.  Does this seem correct to you?  Add @wfbuhl for comment.  I will put some examples up here
https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/4000s/4567

For 5ZoneAirCooled, look at the last column of the ssz around 7-8am.
For RefBldgSmallOfficeNew2004_Chicago, it's just two rows at 22:10 and 22:20 that show diffs for many (all?) of the Des Per 2, Tot Cool Cap columns.

Nigusse commented:
@mjwitte @rraustad @wfbuhl The big diffs in the output for the four test cases are expected due to changes in capacity. When the capacity changes, the HVAC equipment electric energy use changes almost proportionally ( Qrated/COP), the sizing flow rates also have changed that affects the plant electric energy use and hence the facility electric energy. The gas use energy use by supplemental gas coil also changes due to heating capacity reduction (heating capacity takes the size the cooling coil in heat pumps). So, the changes that I see in the four test files is directly consequence of design capacity and flow rate changes. In the other two test files, the results provided by Mike (which I was also able to duplicate the results), there are difference in the sizing but the annual simulation did not shown any diffs in the output report variables and meters.  Since the diff in the design total cooling capacity sizing was not at the peak hours, there was no diffs in total capacity of the coil. Also, note that there is no sizing results diffs for sensible cooling capacity because, the design supply air humidity ratio is used in the sensible capacity sizing calculation.

However, I have a concern of setting the equipment inlet humidity ratio, which is the zone return humidity ratio when no outdoor air, with user specified design supply humidity ratio. This is already done when the heating load is zero, and I extended the same logic for cooling mode in this fix. Otherwise, the original logic (before the fix) whether there is cooling load or not, the the equipment inlet humidity ratio is set to the zone return node value or mixed with the outdoor if there is outdoor air. I do not understand why we did not have the same logic for heating and cooling modes initialization in SizeZoneEquipment routine line# 621-651. Any ways let see alternative solution.

Another approach that I am suggesting in resolving this issue is to change the design supply air humidity ratio of the example file (PipingSystem_Underground_FHX.idf) that results in SHR to be < 1.0 and revert the code change. If we modify the design supply humidity ratio in the zone sizing objects to be comparable to a condition that is expected in zone, in this case Denver, CO weather condition, then the SHR will be < 1.0. I think this is just a bad input assumption; we are requesting a cooling coil to yield humidity ratio of 0.0075 kg/kg when the equipment inlet node humidity ratio was about 0.0045 kg/kg, dry condition. It is for this reason that we are having lower total cooling capacity, or higher sensible cooling capacity.  

At the same time, what should be done in a similar situation when a user specify a design supply air humidity ratio higher than expected at equipment inlet node even if it sizing calculation results in SHR < 1.0? For this defect file if you set "Zone Cooling Design Supply Air Humidity Ratio {kgWater/kgDryAir}" to 0.0055, the SHR will be about 0.86 but the equipment inlet humidity ratio is 0.00452 kg/kg.  Should we allow the "Zone Cooling Design Supply Air Humidity Ratio {kgWater/kgDryAir} to exceed the "FinalZoneSizing( CtrlZoneNum ).DesCoolCoilInHumRat"? This is not captured and that is why we had this bug in the first place.

rraustad commented:
I think this is fine. The only other issue would be to apply a min function such that the coil outlet hum rat is never above the inlet hum rat. Where to do that logically is the question.

mjwitte commented:
@Nigusse @wfbuhl I would not change the example file.  Users can't know for sure what kind of inlet condition they are going to get. My assumption is that the specified cooling design supply humidity ratio is an upper limit.  If the design inlet condition is less, then the design outlet humidity ratio should be set to match the inlet.  But that could be equipment dependent.  Do we have enough information to do this in SizeZoneEquipment?  I think the chilled water coil makes these kinds of adjustments already.

rraustad commented:
I don't think that's the correct place since SizeZoneEquipment is setting up the CalcZone array. This data will be passed to FinalZoneSizing and at that point a check might be made. I can't recall what sizing routine had the problem and what array variables were ultimately used to get to SHR>1. Maybe those variables could be tested where the FinalZoneSizing array is set equal to the Calc array.

Nigusse commented:
For cooling sizing right after this section of the code starting line #  2396 in UpdateZoneSizing routine. we can trap the equipment inlet air humidity not exceed the supply air humidity ratio as follows:

FinalZoneSizing( CtrlZoneNum ).CoolDesHumRat = min( FinalZoneSizing( CurZoneEqNum ).CoolDesHumRat, min( FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak, FinalZoneSizing( CtrlZoneNum ).DesCoolCoilInHumRat ));

This can be added right after the following statement: 
if ( ( FinalZoneSizing( CtrlZoneNum ).TimeStepNumAtCoolMax == 0 || FinalZoneSizing( CtrlZoneNum ).CoolDDNum == 0 ) )

I think such kind of changes would not results in diffs.  We may run into similar large diffs.

mjwitte commented:
@Nigusse This still seems too early. Only the specific piece of equipment knows if it will have any outdoor air at the time of the coil peak.  That's when this check needs to happen.

Nigusse commented:
@mjwitte The proposed change is in UpdateZoneSizing. It is in this routine that the zone final info is passed to terminal equipment sizing:
TermUnitFinalZoneSizing = FinalZoneSizing;

Do you have any suggestion where it should be?

mjwitte commented:
@Nigusse @rraustad For example, fan coil units (and I expect PTAC and the like) might have outdoor air or not.  Only the fan coil unit knows this (if the OA flow rate is zero, it's supposed to size for zero OA, but if the OA flow rate is autosized, it's supposed to size with the design OA flow rate).  Zone sizing can't know what the fan coil is going to do, and there could be other zone equipment that is different.  So, from my understanding, whatever part of the code is making the final calculation of leaving conditions from the coil is where these checks need to happen.  For this particular issue, I don't know if that happens in the WSHP module or in the central sizing routine.

Nigusse commented:
@mjwitte @rraustad This bug fix is stretched more than it should. Any code of this nature may also impacts several equipment types and so the scope of this task needs to account for those as well.

rraustad commented:
I just checked and the central sizing routine does not test for outlet air hum rat < inlet air hum rat.
See RequestSizing and search for SizingType == CoolingCapacitySizing (2 places).

rraustad commented:
Would it be easier to accept this fix and then create an issue for catching this problem once and for all? Both T and w should be tested for cooling. I think testing T for heating is redundant since the coil would wind up with 0 capacity with or without the test..

Nigusse commented:
@mjwitte @rraustad My suggestion is to make code changes to the specific WSHP equipment sizing calculation to address the issue in this particular example file, PipingSystem_Underground_FHX.idf. Then all other equipments needs to resolved in separate bug issue.

mjwitte commented:
@rraustad  I haven't seen an explanation yet why this change is impacting dozens of regression test suite files in the ssz output.  Looking again at this change, some of the statements made earlier today threw me off.

"However, I have a concern of setting the equipment inlet humidity ratio, which is the zone return humidity ratio when no outdoor air, with user specified design supply humidity ratio. This is already done when the heating load is zero, and I extended the same logic for cooling mode in this fix. Otherwise, the original logic (before the fix) whether there is cooling load or not, the the equipment inlet humidity ratio is set to the zone return node value or mixed with the outdoor if there is outdoor air."

This change does none of that.  All this change does is set some things to zero when SysOutputProvided = 0.  All of this discussion about resetting humidity ratios seems irrelevant to this particular fix.

rraustad commented:
See line 621 of ZoneEquipmentManager.cc and see my comment above from 28 days ago. 

The original code set the sizing arrays based on load.
IF heating load THEN
 Initialize sizing arrays with heating data (e.g., outdoor & zone temp/w)
ELSE 
 Initialize sizing arrays with cooling data (e.g., outdoor & zone temp/w)

So now the code says
IF heating load THEN
ELSE IF cooling load THEN
ELSE
  Initialize to 0

So sizing data has changed for files that don't have a cooling load.

Nigusse commented:
I am not sure how I am going to show you the impact of that initialization because several variables are involved but a summary is as follow:  Due to change in the initialization different equipment inlet humidity ratio value is used and that brings about the diffs in capacity, flow rates, ....

When the variables:
 CalcZoneSizing( CurOverallSimDay, ControlledZoneNum ).CoolZoneTemp = 0.0;
CalcZoneSizing( CurOverallSimDay, ControlledZoneNum ).CoolZoneHumRat = 0.0;
 are set to zero when SysOutputProvided=0.

That result in setting FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak = 0.0  and if the ZoneHumRatAtCoolPeak is zero then it is set to CoolDesHumRat.

see existing code line #2382
 if ( FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak > 0.0 ) {
FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak = min( FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak, PsyWFnTdpPb( FinalZoneSizing( CtrlZoneNum ).ZoneTempAtCoolPeak, StdBaroPress, RoutineName ) );
} else {
FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak = ZoneSizing( DDNumF, CtrlZoneNum ).CoolDesHumRat;
}

Again the design supply air humidity ratio is used instead of the equipment inlet humidity ratio. 

FinalZoneSizing( CtrlZoneNum ).DesCoolCoilInTemp = FinalZoneSizing( CtrlZoneNum ).ZoneTempAtCoolPeak;
FinalZoneSizing( CtrlZoneNum ).DesCoolCoilInHumRat = FinalZoneSizing( CtrlZoneNum ).ZoneHumRatAtCoolPeak;
FinalZoneSizing( CtrlZoneNum ).ZoneRetTempAtCoolPeak = FinalZoneSizing( CtrlZoneNum ).ZoneTempAtCoolPeak;

For the four example files, I see the capacity, and supply air flow rate changes, these are changes an being all kind of changes. Does this help?

Nigusse commented:
Sorry last statement is not clear. 
For the four example files, I see the capacity, and supply air flow rate changes, these changes can bring all kind of changes in the energy use of equipments.

Nigusse commented:
@mjwitte @rraustad The cause in diffs for zones with non-zero load (e.g. 5ZoneAirCooled) is due to the resetting (overwriting with zero value) of the zone CalcZoneSizing( CurOverallSimDay, ControlledZoneNum ).CoolZoneHumRat to zero value when the system time step is down shifting. 

When there is relatively small load (stored in ZoneSysEnergyDemand( ActualZoneNum ).RemainingOutputRequired variable) and this load is met in the first detailed time step. Since in the first detailed time step the "SysOutputProvided" is not zero, "_.CoolZoneHumRat" variable is set to zone air node humidity ratio. But by the end of the time step the Remaining Output Required is zero (fully met already), so  "SysOutputProvided" is zero and the "_.CoolZoneHumRat" variable is set to zero in else section of the new code. As the result, with new code, if there is down shifting of the time step and the RemainingOutputRequired is relatively small that it goes to zero by the end of the time step (after down shifting), then the initialization of "_.CoolZoneHumRat" switches from zone air node humidity ratio to a zero value.  A zero "_.CoolZoneHumRat" value was used for calculating the average return air humidity ratio for system sizing. Note that there is supply air flow rate for that particular zone because there is zone load.  Before the fix even if the system is down shifting the initialization always goes to the same section of the code and "*.CoolZoneHumRat" variable is always set to Zone air Node humidity ratio.

This explains why the we are having diffs when there is load. Two steps solution is required to solve this problem.  First, how to avoid overwriting the first initialization with zero value when there is system time step downshifting. Second, what value should we reset the "*.CoolZoneHumRat" variable when it is zero. The second step is required to avoid the issue observed in the defect file. 

Do you have any suggestion how to avoid this switching initialization when there is system time step down shifting? Do you agree setting "*.CoolZoneHumRat" variable to design supply air humidity ration when it is zero?

Nigusse commented:
@mjwitte "For 5ZoneAirCooled, look at the last column of the ssz around 7-8am." I run the simulation with detailed time step, difference was noted in ssz for "DesPer 2:Des Tot Cool Cap" variable where there was system time step downshifting (around 7-8am, and 8:15 pm).

mjwitte commented:
@Nigusse Interesting.  I didn't even realize that downshifting would be active during the sizing calcs, but it must be.  Somehow seems wrong that these values even get revised during the downshift, I wonder how that all feeds into the ssz series of values - they show up in the ssz output as hourly, but how are they stored internally?  I'll poke around a bit.

lgu1234 commented:
@Nigusse @mjwitte I may have a suggestion of "*.CoolZoneHumRat" reset. Can we use values at previous time step to make node values continuous?

Nigusse commented:
The second step of the proposed fix captures when "_.CoolZoneHumRat" is zero and resets it to the supply air design humidity ratio  "_.CoolDesHumRat" in UpdateZoneSizing routine ( under "} else if ( SELECT_CASE_var == DuringDay ) {" statement) as follows:

// check if the CoolZoneHUmRat is zero, then reset it to design supply air humidity ratio as follows:
if ( CalcZoneSizing( ).CoolZoneHumRat == 0.0 ) {
   CalcZoneSizing().CoolZoneHumRat = CalcZoneSizing().CoolDesHumRat
} 

This resolves the SHR > 1 problem and also minimized the SSZ difference observed.  In the case of the 5ZoneAIrCooled.idf test file the maximum diffs noted in the previous run (Mike's regression run) reduced from 13.7% to 1.19%.  I also randomly checked SSZ diffs in my latest regression run  and found out that the diffs in most cases were less than 1.0%. The reason for the diffs reduction is resetting the zero values of CoolZoneHumRat variable with CoolDesHumRat.  Resetting the CoolZoneHumRat variable with CoolDesHumRat is not new approached, it has been already used elsewhere for zero load and no sizing scenarios.

@mjwitte @rraustad @lgu1234 Is this acceptable?

mjwitte commented:
@Nigusse Not sure.  I need to know the context.  I suppose if you want to, push it up, so I can see where it's going.  I'm concerned that both of these changes are impacting general routines in the heart of sizing, so we need to be sure we're not missing any important side effects.

Nigusse commented:
@mjwitte pushed the revised code.

lgu1234 commented:
@mjwitte @Nigusse I have a thought. Since @Nigusse found a problem during system downshifting, is it possible to force sizing calculation using zone time step only? In this way, a sharp load may be avoided and system capacity may be more reasonable using an average value obtained in a zone time step.

rraustad commented:
Here's the code for that if someone wants to test that theory. I think you would only need to add
 !DoingSizing to the if test.

if ( ZoneTempChange > MaxZoneTempDiff && ! KickOffSimulation && !DoingSizing ) {

ManageHVAC()

if ( ZoneTempChange > MaxZoneTempDiff && ! KickOffSimulation ) {
    //determine value of adaptive system time step
    // model how many system timesteps we want in zone timestep
    ZTempTrendsNumSysSteps = int( ZoneTempChange / MaxZoneTempDiff + 1.0 ); // add 1 for truncation
    NumOfSysTimeSteps = min( ZTempTrendsNumSysSteps, LimitNumSysSteps );
    //then determine timestep length for even distribution, protect div by zero
    if ( NumOfSysTimeSteps > 0 ) TimeStepSys = TimeStepZone / NumOfSysTimeSteps;
    TimeStepSys = max( TimeStepSys, MinTimeStepSys );
    UseZoneTimeStepHistory = false;
    ShortenTimeStepSys = true;
} else {
    NumOfSysTimeSteps = 1;
    UseZoneTimeStepHistory = true;
}

rraustad commented:
@mjwitte not sure how we got to multiple PRs. This should be closed.

rraustad commented:
Closed with unmerged commits????

mjwitte commented:
@Nigusse @rraustad This is the original work that touched the sizing routines at a higher level which is now #4893 (perhaps that could use a more general title).  We my want to close this and start over with a new branch, but the discussion here is relevant until we decide whether to change any sizing code or not.

mjwitte commented:
This has been replaced by a different fix, #5012
